plugin/path_checkouts.go

```diff
@@ -5,11 +5,349 @@ import (
+	setName := fieldData.Get("name").(string)
+
+	ttlPeriodRaw, ttlPeriodSent := fieldData.GetOk("ttl")
+	if !ttlPeriodSent {
+		ttlPeriodRaw = 0
+	}
+	requestedTTL := time.Duration(ttlPeriodRaw.(int)) * time.Second
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(`"%s" doesn't exist`, setName), nil
+	}
+
+	// Prepare the check-out we'd like to execute.
+	ttl := set.TTL
+	if ttlPeriodSent {
+		switch {
+		case set.TTL <= 0 && requestedTTL > 0:
+			// The set's TTL is infinite and the caller requested a finite TTL.
+			ttl = requestedTTL
+		case set.TTL > 0 && requestedTTL < set.TTL:
+			// The set's TTL isn't infinite and the caller requested a shorter TTL.
+			ttl = requestedTTL
+		}
+	}
+	newCheckOut := &CheckOut{
+		IsAvailable:         false,
+		BorrowerEntityID:    req.EntityID,
+		BorrowerClientToken: req.ClientToken,
+	}
+
+	// Check out the first service account available.
+	for _, serviceAccountName := range set.ServiceAccountNames {
+		// checkOut is its own function to avoid deferring the lock release inside a loop,
+		// which would cause a memory leak.
+		password, err := b.checkOut(ctx, req.Storage, serviceAccountName, newCheckOut)
+		if err == ErrCheckedOut {
+			continue
+		}
+		respData := map[string]interface{}{
+			"service_account_name": serviceAccountName,
+			"password":             password,
+		}
+		internalData := map[string]interface{}{
+			"service_account_name": serviceAccountName,
+		}
+		resp := b.Backend.Secret(secretAccessKeyType).Response(respData, internalData)
+		resp.Secret.Renewable = true
+		resp.Secret.TTL = ttl
+		resp.Secret.MaxTTL = set.MaxTTL
+		return resp, nil
+	}
+
+	// If we arrived here, it's because we never had a hit for a service account that was available.
+	resp, err := logical.RespondWithStatusCode(&logical.Response{
+		Warnings: []string{"No service accounts available for check-out, please try again later."},
+	}, req, 429)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (b *backend) checkOut(ctx context.Context, storage logical.Storage, serviceAccountName string, newCheckOut *CheckOut) (string, error) {
+	lock := locksutil.LockForKey(b.checkOutLocks, serviceAccountName)
+	lock.Lock()
+	defer lock.Unlock()
+
+	// We've found a service account that we can check out. Let's do it and respond.
+	if err := b.checkOutHandler.CheckOut(ctx, storage, serviceAccountName, newCheckOut); err != nil {
+		return "", err
+	}
+	password, err := retrievePassword(ctx, storage, serviceAccountName)
+	if err != nil {
+		return "", err
+	}
+	return password, nil
+}
+
+func (b *backend) secretAccessKeys() *framework.Secret {
+	return &framework.Secret{
+		Type: secretAccessKeyType,
+		Fields: map[string]*framework.FieldSchema{
+			"service_account_name": {
+				Type:        framework.TypeString,
+				Description: "Access Key",
```

Comment:
Should this be "Service Account", or something other than "Access Key"?

Reply:
Whoops, copy pasta that slipped through.